examples/shop/src/store/bucket.tsx

```diff
@@ -0,0 +1,58 @@
+import React, { createContext, useContext } from 'react';
+import { Product } from '../types';
+
+export type AddedProduct = Product & { count: number };
+
+export type AddedProducts = Map<number, AddedProduct>;
+
+type BucketContextType = {
+    addedProducts: AddedProducts;
+    addProduct: (product: Product) => void;
+    removeProduct: (product: Product) => void;
+};
+
+const BucketContext = createContext<BucketContextType>({} as BucketContextType);
+
+type BucketProviderProps = {
+    children: React.ReactNode;
+};
+
+export function BucketProvider({ children }: BucketProviderProps) {
+    const [addedProducts, setAddedProducts] = React.useState<AddedProducts>(new Map());
+
+    function addProduct(newProduct: Product) {
+        setAddedProducts((addedProducts) => {
+            if (addedProducts.has(newProduct.id)) {
+                const addedProduct = addedProducts.get(newProduct.id) as AddedProduct;
+
+                addedProduct.count += 1;
+            } else {
+                addedProducts.set(newProduct.id, { ...newProduct, count: 1 });
+            }
+
+            return new Map(addedProducts);
+        });
+    }
+
+    function removeProduct(deletedProduct: Product) {
+        if (addedProducts.has(deletedProduct.id)) {
+            setAddedProducts((addedProducts) => {
```

Comment:
Could you rename `addedProduct` to `product`; ('addedProduct' and 'addedProducts' too similar) ?
Please, do not mutate `addedProducts`. This can lead to side effects in the future.

``` tsx

        setAddedProducts((addedProducts) => {
                const product = addedProducts.get(deletedProduct.id);
                if(!product){
                    return addedProducts;
                 }
                 const newAddedProducts = new Map(addedProducts);
                 product.count -= 1;

                if (product.count === 0) {
                    newAddedProducts.delete(deletedProduct.id);
                }

                return newAddedProducts;
            });
```

Reply:
fixed